[WIP] - APPROACH CHECK-IN - v1 progress

davidmarkclements opened the pull request:
initial implementation of the following

* join
* leave
* connect
* ephemeral opt 
* peer-queue integration
* queue lru

tests are passing but anything newly implemented hasn't been checked at all, this PR is intended as a check-in to ensure that the approaches are valid and that peer queue has been correctly understood

key discussion points: 

### leave
the leave method is using discovery caches to find topics - this saves mem but will mean leave is more expensive. I'm assuming for most use cases leave usage would be light, is this assumption correct? If so, we should modify hyperswarm discovery to return a topic for a given key instead of using private apis

### join 
In join I'm aware that discovery can be used to both announce and lookup, and this is more efficient - however we haven't implemented that in guts and I'd rather get something down and tested first. A change to announce+lookup could be made during optimization phase.

Join is making some new assumptions regarding sane option defaults, if the instance is
not ephemeral, `announce` defaults to `true`, otherwise it defaults to `false`. The `lookup`
option defaults to `true` if `announce` is false (same as in v0). 

### ~~hashlru~~

~~I've added https://github.com/dominictarr/hashlru in place of the `Map` in peer queue. 
This is a lightweight and fast LRU, the trade off is that once peer count reaches `LRU_SIZE * 2` then `LRU_SIZE`  amount of peers will be cleared. Is this an acceptable trade off?~~

### peer queue

My understanding of peer queue is that it periodically reprioritises peers based on their connection status. Status is checked via polling, a back-off mechanism using retry count to determine interval duration controls polling. 

It emits a `readable` event when the queue is populated with one or more healthy peers.
Peers can then be shifted in order of priority (which essentially communicates _how_ healthy a peer is).  

Peers are added to the queue when a topic gets a peer event.

My understanding of how to consume peers from the queue is similar to the readable/while approach with node streams, see https://github.com/hyperswarm/network/compare/v1...v1-progress?expand=1#diff-17a266fdf8e53cfbbdd558c38b69cefdR31

Is this the right direction?

The peer queue also has a `remove` method, but I'm not sure where to use this - wouldn't the peer queue be expected to call `remove` on itself once a peer has been banned? Are there other scenarios I'm not seeing where we would need to call remove from `Swarm`?

mafintosh commented on lib/queue.js:
```diff
@@ -1,21 +1,26 @@
 
     this.destroyed = false
-    this._infos = new Map() // TODO: turn into LRU
+    this._infos = lru(cacheSize)
```
i don't this this should be an LRU actually. The queue has a .remove method for removal. No implicit removal should be done.

davidmarkclements replied:
ok, scenarios for explicit removal

* when a peer has been banned?
* when the cache exceeds a memory limit or peer count?
* any others?

mafintosh commented on swarm.js:
```diff
@@ -9,29 +10,89 @@ class Swarm extends EventEmitter {
+      while (peer) {
+        network.connect(peer, onConnect)
+        peer = queue.shift()
+      }
```
We need to only shift from the queue until we hit a 'maxConnections' threshold.

davidmarkclements replied:
got it :+1:

davidmarkclements commented:
* added maxPeers option (default 16)
* queue drains and connects up to a maximum of maxPeers
* added `connectivity` method (umbrella for checking whether peer can bind, join bootstrap nodes, and holepunch)
* converted example to changes

davidmarkclements commented:
dependent on https://github.com/hyperswarm/guts/pull/9

davidmarkclements commented:
* retry and reconnect added

cc @mafintosh - ready for another check in, could probably merge after this iteration to keep PR's bite size

mafintosh commented on swarm.js:
```diff
@@ -1,38 +1,146 @@
+    const onConnect = (info) => (err, socket, isTCP) => {
+      if (err) {
+        this.peers -= 1
+        if (!queue.requeue(info)) queue.requeue(info)
```
what's happening on this line?

davidmarkclements replied:
we're sending it back round the queue in case the error is non-permanent - eventually it'll work itself out via depriotization -> ban right?

davidmarkclements replied:
ohhh wait.. the conditional check is wrong, it's meant to be checking something else I need to remember what it is supposed to be checking. 

Is it correct to requeue at this point thought?

mafintosh replied:
Right to requeue for sure. I actually think you just want to always requeue here.

davidmarkclements replied:
yeah cool ok - whatever I meant to be checking (maybe reconnecting?) I think it should be unnecessary anyway, the queue itself should take care of any conditional logic

mafintosh commented on swarm.js:
```diff
@@ -1,38 +1,146 @@
+        drain()
+        return
+      }
+      if (info.reconnecting) info.reconnect(false)
```
why do we disable reconnects here?

davidmarkclements replied:
I might have misunderstood, but I thought `reconnect(bool)` set whether a peer was reconnecting or not, 
if we've successfully connected then we need to set reconnecting to false

mafintosh replied:
Ah yea I can see the confusion. The RECONNECT flag is used as a capability internally. I.e. should this peer be reconnecting at all? The `info.reconnect(bool)` is the public method to control this.

Something like

```js
swarm.on('connection', function (connection, peerInfo) {
  // do stuff
  if (weDontWantToReconnectToThisPeerWhenItCloses) {
    peerInfo.reconnect(false)
  }
})
```

mafintosh commented on swarm.js:
```diff
@@ -1,38 +1,146 @@
+        this.peers -= 1
+        info.disconnected()
+        info.reconnect(true)
+        if (!queue.requeue(info)) queue.requeue(info)
```
not sure i follow this line again here

davidmarkclements replied:
addressed in prior comment

mafintosh commented on swarm.js:
```diff
@@ -1,38 +1,146 @@
+    }
+    const drain = () => {
+      const info = queue.shift()
+      if (info && this.peers < this.maxPeers) {
```
we want to make move this peers < maxPeers check up before the queue shift

davidmarkclements replied:
of course 🤦‍♂

mafintosh commented:
Awesome stuff @davidmarkclements - I've added a few comments for clarifications.

I'd love to get this merged (after comments are resolved) and then move on to really locking the retry/peer behaivor down with a ton of tests so we have a strong contract on how we think this should behave, wdyt?

davidmarkclements commented:
> wdyt?

agree completely

karissa commented:
super stoked on this! I hear that ara (https://ara.one/) is already using hyperswarm, and other projects will start taking it up soon too :)

davidmarkclements commented:
@mafintosh comments addressed - once approved I'll merge and get into testing
